Srcs/heiftojpeg/heiftojpeg.cpp

```diff
@@ -0,0 +1,183 @@
+            cout << "OINF ";
+            // cout << reader.getPropertyOinf(contextId, property.index);
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::RLOC) {
+            cout << "RLOC ";
+            cout << reader.getPropertyRloc(contextId, property.index).horizontalOffset;
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::TOLS) {
+            cout << "TOLS ";
+            cout << reader.getPropertyTols(contextId, property.index).targetOlsIndex;
+        } else {
+            cout << "unknown ";
+        }
+        cout << "\n";
+    }
+
+    // Always reuse the parameter set from the first tile, sometimes tile 7 or 8 is corrupted
+    HevcImageFileReader::ParameterSetMap parameterSet;
+    reader.getDecoderParameterSets(contextId, tileItemIds.at(0), parameterSet);
+    std::string codeType = reader.getDecoderCodeType(contextId, tileItemIds.at(0));
+    ImageFileReaderInterface::DataVector parametersData;
+    if ((codeType == "hvc1") || (codeType == "lhv1")) {
+        // VPS (HEVC specific)
+        parametersData.insert(parametersData.end(), parameterSet.at("VPS").begin(), parameterSet.at("VPS").end());
+    }
+
+    if ((codeType == "avc1") || (codeType == "hvc1") || (codeType == "lhv1")) {
+        // SPS and PPS
+        parametersData.insert(parametersData.end(), parameterSet.at("SPS").begin(), parameterSet.at("SPS").end());
+        parametersData.insert(parametersData.end(), parameterSet.at("PPS").begin(), parameterSet.at("PPS").end());
+    } else {
+        // No other code types supported
+        // throw FileReaderException(FileReaderException::StatusCode::UNSUPPORTED_CODE_TYPE);
+    }
+
+    ImageFileReaderInterface::DataVector itemDataWithDecoderParameters;
+    ImageFileReaderInterface::DataVector itemData;
+
+    std::vector<Magick::Image> tileImages;
+    for (auto& tileItemId: tileItemIds) {
+        itemDataWithDecoderParameters.clear();
+        itemDataWithDecoderParameters.insert(itemDataWithDecoderParameters.end(), parametersData.begin(), parametersData.end());
+
+        itemData.clear();
+        reader.getItemData(contextId, tileItemId, itemData);
+
+        // +1 comes from skipping first zero after decoder parameters
+        itemDataWithDecoderParameters.insert(itemDataWithDecoderParameters.end(), itemData.begin() + 1, itemData.end());
+
+        Magick::Image image;
+        decodeData(itemDataWithDecoderParameters, &image);
+        tileImages.push_back(image);
+    }
+
+    Magick::Montage montageOptions;
+    montageOptions.tile("8x6");
+    montageOptions.geometry("512x512");
+    std::list<Magick::Image> montage;
+    Magick::montageImages(&montage, tileImages.begin(), tileImages.end(), montageOptions);
+    Magick::Image image = montage.front();
+    image.magick("JPEG");
+    image.crop(Magick::Geometry(gridItem.outputWidth,gridItem.outputHeight));
+    image.rotate(-rotation);
```

Comment:
does this paramater actually use the same enumeration values as IROT and if we rotate ourselves here but leave the orientation field in the exif haven't we created an image that says it needs rotation (based on its exif) but in fact does not and therefore will show up in the wrong orientation when auto-rotated by viewing software like imgix?

Reply:
IROT is in degrees.

Good catch on double-rotation, you are correct. For now I will not have imagemagick rotate them, the EXIF orientations are indeed correct. In the future we may want to rotate the images and then zero the orientation in EXIF.